### PR TITLE
Support for --env-file option for docker-compose

### LINKED ISSUE #6170: --env-file option
This option has been discussed in a [couple](https://github.com/docker/compose/issues/4841#issuecomment-345242740) of [tickets](https://github.com/docker/compose/issues/745) but never had a ticket for itself.

One should be able to execute 

`docker-compose up --env-file=.env.test` (perhaps multiple env files)

to override the usage of `.env` file.

### PR DESCRIPTION
**Pull request by slowr**
Added additional argument (--env-file) for docker-compose to import environment variables from a given file.

Resolves #6170

### CODE REVIEW

**Comment by xeor** [comment]
As noticed in https://github.com/docker/compose/issues/6170#issuecomment-467546158, this PR doesn't allow for multiple environment-files to be loaded at the same time.. Will that be added to a future PR, or was it forgotten about, or not in scope?

Sorry for adding this comment too late. But I'm curious if there are plans for adding multi-env support..

**Comment by ulyssessouza** [comment]
This was considered out of scope for now but we can investigate it. 
And as always... PRs are welcome :smiley_cat:

**Comment by m451** [comment]
Uhm, didn't this make it into 1.24.0? It was commited nearly 2 months ago but I can't find it in the commits for rc1 or rc3 of 1.24.0 and rc2 is missing in the release history.

**Comment by ulyssessouza** [comment]
@m451 You are right. The reason is that the focus of 1.24.0 was around stability in some aspects.
This feature will be present when we release a version based on master.

**Comment by Maher4Ever** [comment]
@ulyssessouza Will this feature be added to the next release?

**Comment by jakedouglas** [comment]
@GammaGames it took me a while to figure out, but it looks like it actually is there. It just doesn't appear in the help output for subcommands, e.g. `docker-compose help up`. `docker-compose help` does show it though, and it appears to work for me.

**Comment by entrptaher** [comment]
Does not work for me on latest version. :(

**Comment by m451** [comment]
Strange. It works for me. What version does your docker-compose report?

**Comment by jcorry** [comment]
This is the ridiculous sort of thing we have to do to compensate for this not being here:
```
start_local:
	cp ./local.env ./.env
	docker-compose up

start_test:
	cp ./test.env ./.env
	docker-compose up
```

**Comment by entrptaher** [comment]
> Strange. It works for me. What version does your docker-compose report?

➜ ~ docker-compose --version
docker-compose version 1.25.0-rc2, build 661ac20e

**This worked** for me,
image

```
ENV_FILE=.env.production.local docker-compose -f docker-compose.prod.yml up --build
```

**These did not work**
- docker-compose up --env-file=
- env $(cat .env) docker-compose up
- BAR=foo docker-compose up

did not work!

**Comment by m451** [comment]
@entrptaher I don't have it in front of me right now but I think if you look at the source it's a global parameter to docker-compose, so as far as I remember I had to use ´docker-compose --env-file=foo up´

**Comment by entrptaher** [comment]
Yes, that's what I said above, `docker-compose up --env-file=foo` did not work. See the `These did not work` points listed on my comment. :)

**Comment by m451** [comment]
@entrptaher `docker-compose up --env-file=foo` is not the same as `docker-compose --env-file=foo up`

**Comment by entrptaher** [comment]
@m451, sure, here you go.
image

As I said, I tested with every combination available on the issue section of this repo.

**Comment by m451** [comment]
actually, it's `docker-compose --env-file foo.env up` without the `=` char. Works for me on `docker-compose version 1.25.0-rc2, build 661ac20e`.

**Comment by entrptaher** [comment]
Okay, let's try that as well, 
image

No success.

**Comment by jkoorts** [comment]
docker-compose config should also with this switch so you can see what will be applied before you hit the red button

**Comment by tiany** [comment]
Failed for me either.

```bash
$docker-compose -v
docker-compose version 1.25.3, build d4d1b42b

$cat docker-compose.yml 
version: '3'
services:
  api:
    image: 'busybox'
    # env_file: .env
    command: /bin/sh -c export

$cat .env 
X=Y

$docker-compose --env-file .env up
Starting dc-test_api_1 ... done
Attaching to dc-test_api_1
api_1  | export HOME='/root'
api_1  | export HOSTNAME='d4d1366336a3'
api_1  | export PATH='/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin'
api_1  | export PWD='/'
api_1  | export SHLVL='1'
dc-test_api_1 exited with code 0
```

**Comment by iraklisg** [comment]
Interestingly, I cannot use the `--env-file` option when running `docker-compose up` in detached mode :confused: 

This works:
```
docker-compose -f docker-compose.testing.yml --env-file .env.testing up my_service
```

This exits gracefully:

```
 docker-compose -f docker-compose.testing.yml --detach --env-file .env.testing up my_service

OUTPUT
--------------------------------------------
Define and run multi-container applications with Docker.

Usage:
  docker-compose [-f <arg>...] [options] [--] [COMMAND] [ARGS...]
  docker-compose -h|--help

Options:
  -f, --file FILE             Specify an alternate compose file
                              (default: docker-compose.yml)
  -p, --project-name NAME     Specify an alternate project name
                              (default: directory name)
  -c, --context NAME          Specify a context name
  --verbose                   Show more output
  --log-level LEVEL           Set log level (DEBUG, INFO, WARNING, ERROR, CRITICAL)
  --no-ansi                   Do not print ANSI control characters
  -v, --version               Print version and exit
  -H, --host HOST             Daemon socket to connect to

  --tls                       Use TLS; implied by --tlsverify
  --tlscacert CA_PATH         Trust certs signed only by this CA
  --tlscert CLIENT_CERT_PATH  Path to TLS certificate file
  --tlskey TLS_KEY_PATH       Path to TLS key file
  --tlsverify                 Use TLS and verify the remote
  --skip-hostname-check       Don't check the daemon's hostname against the
                              name specified in the client certificate
  --project-directory PATH    Specify an alternate working directory
                              (default: the path of the Compose file)
  --compatibility             If set, Compose will attempt to convert keys
                              in v3 files to their non-Swarm equivalent (DEPRECATED)
  --env-file PATH             Specify an alternate environment file

Commands:
  build              Build or rebuild services
  config             Validate and view the Compose file
  create             Create services
  down               Stop and remove containers, networks, images, and volumes
  events             Receive real time events from containers
  exec               Execute a command in a running container
  help               Get help on a command
  images             List images
  kill               Kill containers
  logs               View output from containers
  pause              Pause services
  port               Print the public port for a port binding
  ps                 List containers
  pull               Pull service images
  push               Push service images
  restart            Restart services
  rm                 Remove stopped containers
  run                Run a one-off command
  scale              Set number of containers for a service
  start              Start services
  stop               Stop services
  top                Display the running processes
  unpause            Unpause services
  up                 Create and start containers
  version            Show version information and quit
```

Am I missing something?

**Comment by plueschpruem** [comment]
I hope this is the right place to ask this…
I am building a workflow to use a single docker-compose.yml to manage several similar projects.
I want to solve that with several .env files to use with the --env-file shell option:
`docker-compose --env-file ./Docker_Config/env_files/test1.env -f ./Docker_Config/docker-compose.yml --project-name test1 up -d --build`
Inside the test1.env is this:
`DOCKER_NAME=test1`

File structure:
`root
    Docker_Config
        env_files
                test1.env
         docker-compose.yml`

This works as expected, the app builds, starts, all containers are correctly named via the variabel – so far so dandy.
But when I try to stop the app, I get an error that Docker Desktop could not find the test1.env file.
It seems that Docker is always prepending the full folder of the docker-compose.yml to the .env.
So in my case it tries to open the file at _[…]Docker_Config/Docker_Config/env_files/test1.env_
To confirm I also moved my .yml a level further into another folder – that then also got prepended.
So basically the ONLY possibility to use this new (and frankly very useful) option is to keep all you docker-compose.yml and .env files in the root level. But I guess this is an error, not the desired functionality?

I am on the latest OSX, with Docker Desktop 4.5.0 (74594) / Docker version 20.10.12, build e91ed57 | Docker Compose version v2.2.3

**Comment by entrptaher** [comment]
> I hope this is the right place to ask this… I am building a workflow to use a single docker-compose.yml to manage several similar projects. I want to solve that with several .env files to use with the --env-file shell option: `docker-compose --env-file ./Docker_Config/env_files/test1.env -f ./Docker_Config/docker-compose.yml --project-name test1 up -d --build` Inside the test1.env is this: `DOCKER_NAME=test1`
> 
> File structure: root — Docker_Config —— env_files ——— test1.env —— docker-compose.yml
> 
> This works as expected, the app builds, starts, all containers are correctly named via the variabel – so far so dandy. But when I try to stop the app (with the STOP button in Docker Desktop / Apps), I get an error that Docker Desktop could not find the test1.env file. It seems that Docker is always prepending the full folder of the docker-compose.yml to the .env. So in my case it tries to open the file at _[…]Docker_Config/Docker_Config/env_files/test1.env_ To confirm I also moved my .yml a level further into another folder – that then also got prepended. So basically the ONLY possibility to use this new (and frankly very useful) option is to keep all your docker-compose.yml and .env files in the root level. But I guess this is an error, not the desired functionality?
> 
> Additionally the app console echos some of these on startup: `time="2022-03-12T19:19:25+01:00" level=warning msg="The \"DOCKER_NAME\" variable is not set. Defaulting to a blank string."` Guess that could be connected? Because I only used that var in the docker-config.yml, not in the app itself. It is never passed to the containers and all the functionality the var does provide to the docker-config.yml is working as expectd.
> 
> I am on the latest OSX, with Docker Desktop 4.5.0 (74594) / Docker version 20.10.12, build e91ed57 | Docker Compose version v2.2.3

Can you try this one?

```
alias project="docker-compose --env-file ./Docker_Config/env_files/test1.env -f ./Docker_Config/docker-compose.yml --project-name"

project test1 up --build -d
project down
```

Basically pass all arguments every time.

**Comment by plueschpruem** [comment]
Gives the same error in the Desktop App when I use **stop** instead of **(delete) down** (deleting does still work in the GUI, failed to mention it).
The shell command `project test1 down` works – it lists all containers and takes them down one by one.

So I guess the error lies in the way the desktop app looks for the file maybe?
Bildschirmfoto 2022-03-12 um 20 13 49

This is the error I get

**Comment by XDavidT** [comment]
Is there any option to use more than one `--env-file` flag and let docker compose merge it?
